doc/developer/design/20210630_linearizability.md

```diff
@@ -0,0 +1,84 @@
+and the second system is guaranteed to observe those results (unless they've changed in the meantime).
+
+However, we cannot have a single timestamp number that we guarantee always goes up, because timestamps do not always have the same meaning.
+We have solved the different meaning problem with timelines, which allow joining different sources only if their timestamp number means the same thing.
+For linearizability, we need to have a timestamp number for each timeline.
+
+Add a `HashMap<Timeline, Timestamp>` to track the highest timestamp for a given timeline.
+`sequence_peek`, `sequence_tail`, and `sequence_end_transaction` are the three places timestamps are generated for transactions.
+These three functions will call some new helper function that will either advance or return the most recent timestamp in the HashMap.
+
+Tables are special in that they already provide linearizability.
+Any SQL statement (SELECT or INSERT) that involves a table anywhere uses similar only-goes-up logic to guarantee that INSERTs are always visible to any transaction started after the INSERT.
+The two functions `get_read_ts` and `get_write_ts` provide this.
+Those functions do not need to change because they are sometimes called outside of SQL transactions, and we only need to provide linearizability for user-observed events.
+
+For example,  the `send_builtin_table_updates_at_offset` function calls `get_write_ts` when logging tables are being updated.
+But since that's an internal table update and a user did not observe the write as having happened (which is the case for an INSERT), we don't need to use the HashMap.
+
+This linearizability system can thus be viewed as a layer above the current timestamp determination functions, timelines, and time domains, since it specifically only cares about timestamps that users initiated.
+
+### Wall Time Going Backward
+
+Concerning only to the realtime timeline: the system clock may go backward (due to, say, some time synchronization program like NTP setting the clock).
+This will result in no new data will be produced from (rather, we will not advance the frontiers of) realtime sources.
+SELECTs will continue to function normally, but without any updates until the system clock catches up to what it was.
+
+Tables also have a side effect that if reads and writes are interleaved, the writes will increase the timestamp by at least 1 (if the system time did not increase).
+Thus in worst-case scenarios table writes would be observed, but no new data from realtime sources, until the system clock catches up.
+
+### Restarts
+
+Assume some bad case of materialize shutting down, the system clock going backward a lot, then materialize restarting.
+We care less about the specific timestamp assigned to the data and more that we return any data previously returned.
+On restart, the HashMap must by hydrated with timestamps that will achieve this, even if the timestamps happen to be less than they were before restart.
```

Comment:
I don't quite understand how things work on restart. For a realtime source, the input/ts binding should be persisted so we should start up and timestamp things the same as before, then if we also persist the timelines hashmap it seems like correct behavior would naturally fall out - does that seem right?

System clock regressing sounds like a bigger problem for the source ts/offset bindings, I'm not sure how we deal with that today, if at all.

Reply:
I just had a long talk with Ruchir about source ts bindings and he explained some of what you describe here, so it seems like sources getting themselves to the correct time is fine. We will need to change `determine_timestamp` to consult a source's durability frontier instead of upper frontier, when figuring out a timestamp at which to serve data. If we ever choose a timestamp in advance of any durability frontier in the query, we must wait until the durability frontier passes that timestamp, otherwise we could return different results for the same timestamp after restart. If this is the case, then we don't also need to persist the timelines hashmap because it won't ever exceed the durability frontiers that are already persisted to disk.

I spot checked two places outside the coord crate and both acted correctly if time went backward. I'll check all of them and make sure things look fine, but I was encouraged by what I saw then.